### PR TITLE
feat: add method to set column renderer

### LINKED ISSUE #4941: Add setRenderer(Renderer<T>) to Grid.Column API
### Describe your motivation

This new method would have an impact on the modernization of Vaadin 8 to Flow 24 applications where the [setRenderer(Renderer<? super V> renderer)](https://vaadin.com/api/framework/8.18.0/com/vaadin/ui/Grid.Column.html#setRenderer-com.vaadin.ui.renderers.Renderer-) and/or [setRenderer(ValueProvider<V,P> presentationProvider, Renderer<? super P> renderer)](https://vaadin.com/api/framework/8.18.0/com/vaadin/ui/Grid.Column.html#setRenderer-com.vaadin.data.ValueProvider-com.vaadin.ui.renderers.Renderer-) were frequently used.
Also for new developments it would provide a way to set or change the renderer after the call to Grid.addColumn specially when the signature [addColumn(String propertyName)](https://vaadin.com/api/platform/24.0.4/com/vaadin/flow/component/grid/Grid.html#addColumn(java.lang.String)) without a Renderer is used. 

### Describe the solution you'd like

New Grid.Column methods available:

- setRenderer(Renderer<? super V> renderer)
- setRenderer(ValueProvider<V,P> presentationProvider, Renderer<? super P> renderer)

### Describe alternatives you've considered

The approach seems to be straightforward so no alternatives have been considered.

### Additional context

Not sure if it will work out of the box, because the current implementation assumes setup on first roundtrip. Further testing is required.

### PR DESCRIPTION
**Pull request by sosa-vaadin**
## Description

This PR introduces two methods to change the renderer used to render the contents of a column in a `Grid`.

Fixes #4941 

## Type of change

- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.
- [ ] I have not completed some of the steps above and my pull request can be closed immediately.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 1]
```diff
@@ -527,6 +527,50 @@ public Renderer<T> getRenderer() {
...
+         *            be {@code null}
+         *
+         * @since 24.1
+         */
+        public void setRenderer(Renderer<T> renderer) {
+            this.renderer = Objects.requireNonNull(renderer);
+
+            destroyDataGenerators();
+
+            rendering = renderer.render(getElement(), (KeyMapper<T>) getGrid()
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:544]: Renderings are not properly cleaned up at the moment. There is a `Rendering.getRegistration()`, which should be called to dispose the old renderer. See `VirtualList.setRenderer` for an example.

---

Semi-related, `ComponentRenderer` is not properly cleaned up when removing its registration, the container element that it adds is not removed from the Flow `Column`s element.

Changing the integration test to:
```
        NativeButton swapRenderersButton = new NativeButton(
                "Swap price column renderer",
                clickEvent -> {
                    priceColumn.setRenderer(new ComponentRenderer<>(Span::new, (span, item) -> {
                        span.setText("$" + item.getPrice());
                    }));
                    int virtualChildrenSize = priceColumn.getElement().getNode().getFeature(VirtualChildrenList.class).size();
                    System.out.println("Column virtual children size: " + virtualChildrenSize);
                });
```

Logs on sub-sequent clicks:
```
Virtual children size: 2
Virtual children size: 3
Virtual children size: 4
Virtual children size: 5
Virtual children size: 6
Virtual children size: 7
```

While only indirectly related to this PR, I feel that should be resolved somehow, though ideally in a separate PR.

**Review by sissbruecker — CHANGES_REQUESTED** [review]

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersIT.java [hunk 2]
```diff
@@ -64,6 +63,50 @@ public void basicRenderers_rowsAreRenderedAsExpected() {
...
+        assertRendereredContent("<span style=\"color: red\">US$73.10</span>",
+                TestHelper.stripComments(grid.getCell(0, 1).getInnerHTML()));
+        assertRendereredContent("<span style=\"color: blue\">US$24.05</span>",
+                TestHelper.stripComments(grid.getCell(1, 1).getInnerHTML()));
+    }
+
+    @Test
+    public void swapRendererWithValueProvider() {
+        var priceSorter = grid.getHeaderCell(1).$("vaadin-grid-sorter").first();
+        // content should not get sorted with now value provider
```
[the remark this answers is not in the document]
Comment (sosa-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersIT.java:83]: There was indeed a typo.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 3]
```diff
@@ -527,6 +527,50 @@ public Renderer<T> getRenderer() {
...
+        /**
+         * Switch the renderer used for this column.
+         *
+         * @param renderer
+         *            the new renderer to be used for this column, should never
+         *            be {@code null}
+         *
+         * @since 24.1
+         */
+        public void setRenderer(Renderer<T> renderer) {
```
Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:539]: If the renderer is set _after_ calling `column.setEditorComponent(field)`, the editor doesn't render:

Example (used `GridViewEditorPage.java`):
```java
nameColumn.setEditorComponent(field);

// If I change the renderer _after_ setting the editor component, the
// editor for the column is rendered as "null"
nameColumn.setRenderer(LitRenderer.<Person>of("<b>${item.firstName}</b>")
                .withProperty("firstName", item -> item.getFirstName()));
```

https://user-images.githubusercontent.com/1222264/233973974-2a7c7d5a-19a2-49ed-a40b-c155a3e3a84b.mp4

Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:539]: Seems related to calling `destroyDataGenerators` when setting the renderer, which also removes the editor data generator.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 4]
```diff
@@ -527,6 +527,50 @@ public Renderer<T> getRenderer() {
...
+         * @param valueProvider
+         *            the value provider, used for sorting
+         * @param renderer
+         *            the new renderer to be used for this column, should never
+         *            be {@code null}
+         *
+         * @since 24.1
+         */
+        public <V extends Comparable<? super V>> void setRenderer(
+                ValueProvider<T, V> valueProvider, Renderer<T> renderer) {
```
[the remark this answers is not in the document]
Comment (web-padawan) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:588]: @sosa-vaadin Hi, has this been addressed? If so, please add a comment indicating a change. Thank you.

**Review by web-padawan — APPROVED** [review]
Overall LGTM, but another review is necessary before merging.

Please submit a follow-up PR to address concerns raised by @sissbruecker.

**Review by sissbruecker — CHANGES_REQUESTED** [review]
Blocking this again: The new method for setting a renderer together with a value provider does something different than the framework version of the method. See https://github.com/vaadin/flow-components/issues/4941#issuecomment-1531818885.

**Comment by mgarciavaadin** [comment]
Hi all, let's split the issue for each one of the signatures so we can progress separately as it seems the second signature (with value provider) requires an investigation.

**Comment by sosa-vaadin** [comment]
The second signature of the method was removed in order to address it in a different PR. I've also made the method return the column instance so that chaining method calls is possible, just like in the V8 version.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java [hunk 5]
```diff
@@ -74,8 +106,51 @@ private void createBasicRenderers() {
             dataProvider.refreshAll();
         })).setWidth("100px").setFlexGrow(0);
 
+        Collection<Button> editButtons = Collections
+                .newSetFromMap(new WeakHashMap<>());
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java:110]: Disabling the buttons doesn't seem relevant for the test setup.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java [hunk 6]
```diff
@@ -74,8 +106,51 @@ private void createBasicRenderers() {
...
+                .forEach(button -> button.setEnabled(!editor.isOpen())));
+
+        Button save = new Button("Save", e -> editor.save());
+
+        Button cancel = new Button("Cancel", e -> editor.cancel());
+
+        // Add a keypress listener that listens for an escape key up event.
+        // Note! some browsers return key as Escape and some as Esc
+        grid.getElement().addEventListener("keyup", event -> editor.cancel())
+                .setFilter("event.key === 'Escape' || event.key === 'Esc'");
```
[the remark this answers is not in the document]
Comment (sosa-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java:134]: I will remove this part.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java (hunk 5 above)
Comment (sosa-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java:110]: It isn't, but it avoids multiple rows to be edited at once. It was just copied from `GridViewEditorPage`. I can remove the disabling logic if you find it necessary.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java [hunk 7]
```diff
@@ -43,11 +54,32 @@ private void createBasicRenderers() {
...
+        Binder<Item> binder = new Binder<>(Item.class);
+        Editor<Item> editor = grid.getEditor();
+        editor.setBinder(binder);
+        editor.setBuffered(true);
+
+        TextField nameField = new TextField();
+        binder.forField(nameField).bind("name");
+        nameColumn.setEditorComponent(nameField);
+        nameColumn.setRenderer(LitRenderer.<Item> of("<b>${item.name}</b>")
+                .withProperty("name", Item::getName));
```
[the remark this answers is not in the document]
Comment (sosa-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java:69]: I added a comment explaining the need for these lines to stay in this order.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java (hunk 5 above)
Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java:113]: Let's remove everything that isn't necessary.

Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridViewUsingRenderersPage.java:113]: @sosa-vaadin Doesn't look like this is resolved yet?

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**